linkerd/app/outbound/src/target.rs

```diff
@@ -283,13 +277,38 @@ impl<P> Into<metrics::EndpointLabels> for &'_ Endpoint<P> {
 impl<P: std::hash::Hash> std::hash::Hash for Endpoint<P> {
     fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
         self.addr.hash(state);
-        self.identity.hash(state);
+        self.tls.hash(state);
         self.concrete.resolve.hash(state);
         self.concrete.logical.orig_dst.hash(state);
         self.concrete.logical.protocol.hash(state);
     }
 }
 
+impl EndpointFromMetadata {
+    fn client_tls(metadata: &Metadata) -> tls::ConditionalClientTls {
+        // If we're transporting an opaque protocol OR we're communicating with
+        // a gateway, then set an ALPN value indicating support for a transport
+        // header.
+        let alpn = if metadata.opaque_transport_port().is_some()
+            || metadata.authority_override().is_some()
+        {
+            Some(tls::client::AlpnProtocols(vec![
+                transport_header::PROTOCOL.into()
+            ]))
+        } else {
+            None
+        };
```

Comment:
nit: it looks like we construct this vec regardless of whether or not the metadata has an identity, but if it doesn't have one, we always return `Conditional::None`. Maybe we should refactor this so that the `AlpnProtocols` is only constructed inside the map closure so we don't always build the vec?

probably not a huge deal.

Reply:
note that both of these cases practically require identity to be set (gateway and opaque transport) anyway; but i've fixed this for clarity